pkg/loop/internal/capabilities.go

```diff
@@ -0,0 +1,587 @@
+}
+
+type CallbackCapability interface {
+	capabilities.CallbackExecutable
+	capabilities.BaseCapability
+}
+
+func NewCallbackCapabilityClient(brokerExt *brokerExt, conn *grpc.ClientConn) CallbackCapability {
+	return &CallbackCapabilityClient{
+		callbackExecutableClient: newCallbackExecutableClient(brokerExt, conn),
+		baseCapabilityClient:     newBaseCapabilityClient(brokerExt, conn),
+	}
+}
+
+func RegisterCallbackCapabilityServer(server *grpc.Server, broker Broker, brokerCfg BrokerConfig, impl CallbackCapability) error {
+	bext := &brokerExt{
+		BrokerConfig: brokerCfg,
+		broker:       broker,
+	}
+	pb.RegisterCallbackExecutableServer(server, newCallbackExecutableServer(bext, impl))
+	pb.RegisterBaseCapabilityServer(server, newBaseCapabilityServer(impl))
+	return nil
+}
+
+func RegisterTriggerCapabilityServer(server *grpc.Server, broker Broker, brokerCfg BrokerConfig, impl capabilities.TriggerCapability) error {
+	bext := &brokerExt{
+		BrokerConfig: brokerCfg,
+		broker:       broker,
+	}
+	pb.RegisterTriggerExecutableServer(server, newTriggerExecutableServer(bext, impl))
+	pb.RegisterBaseCapabilityServer(server, newBaseCapabilityServer(impl))
+	return nil
+}
+
+type baseCapabilityServer struct {
+	pb.UnimplementedBaseCapabilityServer
+
+	impl capabilities.BaseCapability
+}
+
+func newBaseCapabilityServer(impl capabilities.BaseCapability) *baseCapabilityServer {
+	return &baseCapabilityServer{impl: impl}
+}
+
+var _ pb.BaseCapabilityServer = (*baseCapabilityServer)(nil)
+
+func (c *baseCapabilityServer) Info(ctx context.Context, request *emptypb.Empty) (*pb.CapabilityInfoReply, error) {
+	info, err := c.impl.Info(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var ct pb.CapabilityType
+	switch info.CapabilityType {
+	case capabilities.CapabilityTypeTrigger:
+		ct = pb.CapabilityType_CAPABILITY_TYPE_TRIGGER
+	case capabilities.CapabilityTypeAction:
+		ct = pb.CapabilityType_CAPABILITY_TYPE_ACTION
+	case capabilities.CapabilityTypeConsensus:
+		ct = pb.CapabilityType_CAPABILITY_TYPE_CONSENSUS
+	case capabilities.CapabilityTypeTarget:
+		ct = pb.CapabilityType_CAPABILITY_TYPE_TARGET
+	}
+
+	return &pb.CapabilityInfoReply{
+		Id:             info.ID,
+		CapabilityType: ct,
+		Description:    info.Description,
+		Version:        info.Version,
+	}, nil
+}
+
+type baseCapabilityClient struct {
+	grpc pb.BaseCapabilityClient
+	*brokerExt
+}
+
+var _ capabilities.BaseCapability = (*baseCapabilityClient)(nil)
+
+func newBaseCapabilityClient(brokerExt *brokerExt, conn *grpc.ClientConn) *baseCapabilityClient {
+	return &baseCapabilityClient{grpc: pb.NewBaseCapabilityClient(conn), brokerExt: brokerExt}
+}
+
+func (c *baseCapabilityClient) Info(ctx context.Context) (capabilities.CapabilityInfo, error) {
+	resp, err := c.grpc.Info(ctx, &emptypb.Empty{})
+	if err != nil {
+		return capabilities.CapabilityInfo{}, err
+	}
+
+	var ct capabilities.CapabilityType
+	switch resp.CapabilityType {
+	case pb.CapabilityType_CAPABILITY_TYPE_TRIGGER:
```

Comment:
Is this the idiomatic way to name these? I guess we shouldn't favor Go over other languages in our protos?

Reply:
Yes, unfortunately! This is because proto enum member fields are globally named, so they recommend prepending the enum type in front of it. I could just add it as "TRIGGER", but that would get awkward with the first element, which by convention is called "UNKNOWN" and would cause issues for other enums in the package.

Reply:
I agree it's really awkward. I could reduce the stuttering less by prepending each member name with an abbreviation, eg. CT_TRIGGER/CT_ACTION/CT_TARGET etc?

Reply:
Could we put some aliases/consts in front of them, with cleaner names?
```go
const CapabilityTypeTrigger = pb.CapabilityType_CAPABILITY_TYPE_TRIGGER
```
Or maybe even from a file in the same package that is generated, so we can do `pb.CapabilityTypeTrigger` here?
```go
const CapabilityTypeTrigger = CapabilityType_CAPABILITY_TYPE_TRIGGER
```